src/meta/src/hummock/manager/mod.rs

```diff
@@ -2459,6 +2404,235 @@ where
+        )>,
+    ) -> (JoinHandle<()>, Sender<()>) {
+        use futures::StreamExt;
+        let mut compactor_request_streams = FuturesUnordered::new();
+        let (shutdown_tx, shutdown_rx) = tokio::sync::oneshot::channel();
+        let shutdown_rx_shared = shutdown_rx.shared();
+        let mut compaction_selectors = init_selectors();
+
+        let join_handle = tokio::spawn(async move {
+            let push_stream =
+                |context_id: u32,
+                 stream: Streaming<SubscribeCompactionEventRequest>,
+                 compactor_request_streams: &mut FuturesUnordered<_>| {
+                    let future = stream
+                        .into_future()
+                        .map(move |stream_future| (context_id, stream_future));
+
+                    compactor_request_streams.push(future);
+                };
+
+            loop {
+                let shutdown_rx_shared = shutdown_rx_shared.clone();
+
+                tokio::select! {
+                    _ = shutdown_rx_shared => {
+                        return;
+                    },
+
+                    compactor_stream = compactor_streams_change_rx.recv() => {
+                        if let Some((context_id, stream)) = compactor_stream {
+                                push_stream(context_id, stream, &mut compactor_request_streams);
+                        }
+                    },
+
+                    result = pending_on_none(compactor_request_streams.next()) => {
+                        let mut compactor_alive = true;
+                        let (context_id, compactor_stream_req) = result;
+                        let (event, stream) = match compactor_stream_req {
+                            (Some(Ok(req)), stream) => {
+                                (req.event.unwrap(), stream)
+                            }
+                            _ => {continue;}
+                        };
+
+                        match event {
+                            RequestEvent::PullTask(PullTask {
+                                pull_task_count,
+                            }) => {
+                                assert_ne!(0, pull_task_count);
+                                if let Some(compactor) = hummock_manager.compactor_manager.get_compactor(context_id) {
+                                    use rand::prelude::SliceRandom;
+                                    use rand::thread_rng;
+                                    let mut compaction_group_ids = hummock_manager.compaction_group_ids().await;
+                                    compaction_group_ids.shuffle(&mut thread_rng());
+
+                                    let (group, task_type) = {
+                                        let (mut group,mut task_type) = (None, None);
+                                        for cg_id in compaction_group_ids {
+                                            if let Some(pick_type) = hummock_manager.compaction_state.auto_pick_type(cg_id) {
```

Comment:
IIUC, if there is nothing to pick for the randomly chosen compaction group, no compact task will be generated even though compaction can be triggered in the non-chosen group. Would it be better to round robin all compaction groups until chosen instead of picking a group randomly?

Reply:
It just shuffle the compaction_group_ids_vec randomly, to avoid starvation. 

The logic here still traverses all cgs. In other words, any cg can be triggered, we can execute it